dandiapi/api/services/embargo/__init__.py

```diff
@@ -33,28 +38,65 @@ def _delete_asset_blob_tags(client: S3Client, blob: str):
 
 # NOTE: In testing this took ~2 minutes for 100,000 files
 def _remove_dandiset_asset_blob_embargo_tags(dandiset: Dandiset):
-    # First we need to generate a CSV manifest containing all asset blobs that need to be untaged
-    embargoed_asset_blobs = AssetBlob.objects.filter(
-        embargoed=True, assets__versions__dandiset=dandiset
-    ).values_list('blob', flat=True)
+    embargoed_asset_blobs = list(
+        AssetBlob.objects.filter(embargoed=True, assets__versions__dandiset=dandiset).values_list(
+            'blob', flat=True
+        )
+    )
 
     client = get_boto_client(config=Config(max_pool_connections=100))
     with ThreadPoolExecutor(max_workers=100) as e:
-        for blob in embargoed_asset_blobs:
+        futures = [
             e.submit(_delete_asset_blob_tags, client=client, blob=blob)
+            for blob in embargoed_asset_blobs
+        ]
+
+    # Check if any failed and raise exception if so
+    failed = [
+        blob for i, blob in enumerate(embargoed_asset_blobs) if futures[i].exception() is not None
+    ]
+    if failed:
+        raise AssetTagRemovalError('Some blobs failed to remove tags', blobs=failed)
 
 
 @transaction.atomic()
-def _unembargo_dandiset(dandiset: Dandiset):
-    # NOTE: Before proceeding, all asset blobs must have their embargoed tags removed in s3
+def _unembargo_dandiset(ds: Dandiset):
+    logger.info('Unembargoing Dandiset %s', ds.identifier)
+    logger.info('\t%s assets', ds.draft_version.assets.count())
+
+    if ds.embargo_status != Dandiset.EmbargoStatus.UNEMBARGOING:
+        raise DandiError(
+            message=f'Expected dandiset state {Dandiset.EmbargoStatus.UNEMBARGOING}, found {ds.embargo_status}',  # noqa: E501
+            http_status_code=500,
```

Comment:
Is this really a 5xx error? What if a user hits the unembargo endpoint for an ineligible Dandiset? That wouldn't be a server error.

Reply:
It'a a 5xx error here because this is an internal function, and is only intended to be called on dandisets for which the embargo status is `UNEMBARGOING`. The workflow is:

1. User calls `unembargo` endpoint, which calls `kickoff_dandiset_unembargo`
2. `kickoff_dandiset_unembargo` updates that dandiset's status to `UNEMBARGOING` in a transaction, which once completed, kicks off `unembargo_dandiset_task`
3. `unembargo_dandiset_task`, once picked up by a celery worker, calls this function, `unembargo_dandiset`

So at step 3, if the dandiset were to _not_ be in the `UNEMBARGOING` state, that would be an internal error.